Heh, I was all ready to show you 10m of good runs and hit this error:

```
--- FAIL: TestProgressWithDownNode (1.41s)
        <autogenerated>:32: storage/client_raft_test.go:812: condition failed to evaluate within 1s: storage/client_raft_test.go:809: expected [5 5 5], got [5 5 0]
```

Does that ring a bell?

--- a/GLOCKFILE
+++ b/GLOCKFILE
@@ -50,7 +50,7 @@ github.com/spf13/pflag 7f60f83a2c81bc3c3c0d5297f61ddfa68da9d3b7
 github.com/tebeka/go2xunit 1dc98d1bb46ddc80cede27ce57b0376948005a31
 github.com/termie/go-shutil bcacb06fecaeec8dc42af03c87c6949f4a05c74c
 golang.org/x/crypto f18420efc3b4f8e9f3d51f6bd2476e92c46260e9
-golang.org/x/net 0cb26f788dd4625d1956c6fd97ffc4c90669d129
+golang.org/x/net 0ab009005dc16437045369e823a35d1af6232b69
 golang.org/x/text cf4986612c83df6c55578ba198316d1684a9a287
 golang.org/x/tools 6e2f52e572ea836c54886d4856872cf239800498
 gopkg.in/yaml.v1 9f9df34309c04878acc86042b16630b0f696e1de

--- a/cli/start.go
+++ b/cli/start.go
@@ -86,10 +86,6 @@ func runInit(_ *cobra.Command, _ []string) error {
 	stopper := stop.NewStopper()
 	defer stopper.Stop()
 
-	return initCluster(stopper)
-}
-
-func initCluster(stopper *stop.Stopper) error {
 	// Default user for servers.
 	context.User = security.NodeUser
 
@@ -100,6 +96,10 @@ func initCluster(stopper *stop.Stopper) error {
 		return util.Errorf("cannot bootstrap to more than one store")
 	}
 
+	return initCluster(stopper)
+}
+
+func initCluster(stopper *stop.Stopper) error {
 	// Generate a new UUID for cluster ID and bootstrap the cluster.
 	clusterID := uuid.NewUUID4().String()
 	if _, err := server.BootstrapCluster(clusterID, context.Engines, stopper); err != nil {
@@ -150,20 +150,23 @@ func runStart(_ *cobra.Command, _ []string) error {
 		// TODO(marc): set this in the zones table when we have an entry
 		// for the default cluster-wide zone config.
 		config.DefaultZoneConfig.ReplicaAttrs = []roachpb.Attributes{{}}
+		context.Stores = "mem=1073741824"
+		context.GossipBootstrap = server.SelfGossipAddr
 	}
 
 	stopper := stop.NewStopper()
-	if context.EphemeralSingleNode {
-		context.Stores = "mem=1073741824"
-		context.GossipBootstrap = server.SelfGossipAddr
+	if err := context.InitStores(stopper); err != nil {
+		return util.Errorf("failed to initialize stores: %s", err)
+	}
 
-		if err := initCluster(stopper); err != nil {
+	if context.EphemeralSingleNode {
+		// A separate stopper for bootstrapping so that we can properly shutdown
+		// all of the stores.
+		initStopper := stop.NewStopper()
+		if err := initCluster(initStopper); err != nil {
 			return err
 		}
-	} else {
-		if err := context.InitStores(stopper); err != nil {
-			return util.Errorf("failed to initialize stores: %s", err)
-		}
+		initStopper.Stop()
 	}
 
 	if err := context.InitNode(); err != nil {

--- a/storage/client_test.go
+++ b/storage/client_test.go
@@ -163,12 +163,6 @@ type multiTestContext struct {
 	// test individually. clientStopper is for 'db', transportStopper is
 	// for 'transport', and the 'stoppers' slice corresponds to the
 	// 'stores'.
-	// TODO(bdarnell): scannerStopper is a hack. The scanner is the one
-	// source of asynchronous operations that the test has no direct
-	// control over, and it is prone to an infinite retry loop during
-	// shutdown. When we have fixed our retry configurations (#2500),
-	// we should be able to remove scannerStopper.
-	scannerStopper     *stop.Stopper
 	clientStopper      *stop.Stopper
 	transportStopper   *stop.Stopper
 	engineStoppers     []*stop.Stopper
@@ -201,9 +195,6 @@ func (m *multiTestContext) Start(t *testing.T, numStores int) {
 		rpcContext := rpc.NewContext(&base.Context{}, m.clock, nil)
 		m.gossip = gossip.New(rpcContext, gossip.TestBootstrap)
 		m.gossip.SetNodeID(math.MaxInt32)
-	}
-	if m.scannerStopper == nil {
-		m.scannerStopper = stop.NewStopper()
 	}
 	if m.clientStopper == nil {
 		m.clientStopper = stop.NewStopper()
@@ -253,7 +244,7 @@ func (m *multiTestContext) Stop() {
 	go func() {
 		m.mu.RLock()
 		defer m.mu.RUnlock()
-		stoppers := append([]*stop.Stopper{m.scannerStopper, m.clientStopper, m.transportStopper},
+		stoppers := append([]*stop.Stopper{m.clientStopper, m.transportStopper},
 			m.stoppers...)
 		// Quiesce all the stoppers so that we can stop all stoppers in unison.
 		for _, s := range stoppers {
@@ -370,7 +361,6 @@ func (m *multiTestContext) makeContext(i int) storage.StoreContext {
 	ctx.StorePool = m.storePool
 	ctx.Transport = m.transport
 	ctx.EventFeed = m.feed
-	ctx.ScannerStopper = m.scannerStopper
 	return ctx
 }
 

--- a/storage/queue.go
+++ b/storage/queue.go
@@ -23,6 +23,8 @@ import (
 	"sync/atomic"
 	"time"
 
+	"golang.org/x/net/trace"
+
 	"github.com/cockroachdb/cockroach/config"
 	"github.com/cockroachdb/cockroach/gossip"
 	"github.com/cockroachdb/cockroach/roachpb"
@@ -126,6 +128,8 @@ type baseQueue struct {
 	replicas    map[roachpb.RangeID]*replicaItem // Map from RangeID to replicaItem (for updating priority)
 	// Some tests in this package disable queues.
 	disabled int32 // updated atomically
+
+	eventLog trace.EventLog
 }
 
 // makeBaseQueue returns a new instance of baseQueue with the
@@ -143,6 +147,7 @@ func makeBaseQueue(name string, impl queueImpl, gossip *gossip.Gossip, maxSize i
 		incoming: make(chan struct{}, 1),
 		Locker:   new(sync.Mutex),
 		replicas: map[roachpb.RangeID]*replicaItem{},
+		eventLog: trace.NewEventLog("queue", name),
 	}
 }
 
@@ -160,6 +165,10 @@ func (bq *baseQueue) SetDisabled(disabled bool) {
 	} else {
 		atomic.StoreInt32(&bq.disabled, 0)
 	}
+}
+
+func (bq *baseQueue) Close() {
+	bq.eventLog.Finish()
 }
 
 // Start launches a goroutine to process entries in the queue. The
@@ -189,8 +198,9 @@ func (bq *baseQueue) MaybeAdd(repl *Replica, now roachpb.Timestamp) {
 	cfg := bq.gossip.GetSystemConfig()
 	if cfg == nil {
 		if log.V(1) {
-			log.Infof("no system config available. skipping...")
+			log.Infof("no system config available. skipping")
 		}
+		bq.eventLog.Printf("no system config available. skipping")
 		return
 	}
 
@@ -199,16 +209,19 @@ func (bq *baseQueue) MaybeAdd(repl *Replica, now roachpb.Timestamp) {
 		// Range needs to be split due to zone configs, but queue does
 		// not accept unsplit ranges.
 		if log.V(3) {
-			log.Infof("range %s needs to be split; not adding", repl)
+			log.Infof("%s needs to be split; not adding", repl)
 		}
+		bq.eventLog.Printf("%s: split needed; not adding", repl)
 		return
 	}
 
 	bq.Lock()
 	defer bq.Unlock()
 	should, priority := bq.impl.shouldQueue(now, repl, cfg)
-	if err := bq.addInternal(repl, should, priority); err != nil && log.V(3) {
-		log.Infof("couldn't add %s to queue %s: %s", repl, bq.name, err)
+	if err := bq.addInternal(repl, should, priority); err != nil {
+		if log.V(3) {
+			log.Infof("unable to add %s to queue %s: %s", repl, bq.name, err)
+		}
 	}
 }
 
@@ -218,6 +231,7 @@ func (bq *baseQueue) MaybeAdd(repl *Replica, now roachpb.Timestamp) {
 // added.
 func (bq *baseQueue) addInternal(repl *Replica, should bool, priority float64) error {
 	if atomic.LoadInt32(&bq.disabled) == 1 {
+		bq.eventLog.Printf("queue disabled")
 		return errQueueDisabled
 	}
 
@@ -226,18 +240,23 @@ func (bq *baseQueue) addInternal(repl *Replica, should bool, priority float64) e
 	item, ok := bq.replicas[rangeID]
 	if !should {
 		if ok {
+			bq.eventLog.Printf("%s: removing", item.value)
 			bq.remove(item.index)
 		}
 		return errReplicaNotAddable
 	} else if ok {
+		if item.priority != priority {
+			bq.eventLog.Printf("%s: updating priority: %0.3f -> %0.3f", repl, item.priority, priority)
+		}
 		// Replica has already been added; update priority.
 		bq.priorityQ.update(item, priority)
 		return nil
 	}
 
 	if log.V(3) {
-		log.Infof("adding replica %s to %s queue", repl, bq.name)
+		log.Infof("adding %s to %s queue", repl, bq.name)
 	}
+	bq.eventLog.Printf("%s: adding: priority=%0.3f", repl, priority)
 	item = &replicaItem{value: repl, priority: priority}
 	heap.Push(&bq.priorityQ, item)
 	bq.replicas[rangeID] = item
@@ -262,8 +281,9 @@ func (bq *baseQueue) MaybeRemove(repl *Replica) {
 	defer bq.Unlock()
 	if item, ok := bq.replicas[repl.Desc().RangeID]; ok {
 		if log.V(3) {
-			log.Infof("removing replica %s from %s queue", item.value, bq.name)
+			log.Infof("removing %s from %s queue", item.value, bq.name)
 		}
+		bq.eventLog.Printf("%s: removing", item.value)
 		bq.remove(item.index)
 	}
 }
@@ -273,7 +293,6 @@ func (bq *baseQueue) MaybeRemove(repl *Replica) {
 //
 // TODO(spencer): current load should factor into replica processing timer.
 func (bq *baseQueue) processLoop(clock *hlc.Clock, stopper *stop.Stopper) {
-
 	stopper.RunWorker(func() {
 		// nextTime is initially nil; we don't start any timers until the queue
 		// becomes non-empty.
@@ -340,7 +359,7 @@ func (bq *baseQueue) processReplica(repl *Replica, clock *hlc.Clock) {
 	// Load the system config.
 	cfg := bq.gossip.GetSystemConfig()
 	if cfg == nil {
-		log.Infof("no system config available. skipping...")
+		log.Infof("no system config available. skipping")
 		return
 	}
 
@@ -349,12 +368,10 @@ func (bq *baseQueue) processReplica(repl *Replica, clock *hlc.Clock) {
 		// Range needs to be split due to zone configs, but queue does
 		// not accept unsplit ranges.
 		if log.V(3) {
-			log.Infof("range %s needs to be split; skipping processing", repl)
+			log.Infof("%s needs to be split; skipping", repl)
 		}
+		bq.eventLog.Printf("%s: split needed; skipping", repl)
 		return
-	}
-	if log.V(3) {
-		log.Infof("processing replica %s from %s queue...", repl, bq.name)
 	}
 
 	// If the queue requires a replica to have the range leader lease in
@@ -364,15 +381,26 @@ func (bq *baseQueue) processReplica(repl *Replica, clock *hlc.Clock) {
 		// Create a "fake" get request in order to invoke redirectOnOrAcquireLease.
 		if err := repl.redirectOnOrAcquireLeaderLease(nil /* Trace */, now); err != nil {
 			if log.V(3) {
-				log.Infof("this replica of %s could not acquire leader lease; skipping...", repl)
+				log.Infof("this replica of %s could not acquire leader lease; skipping", repl)
 			}
+			bq.eventLog.Printf("%s: could not acquire leader lease; skipping", repl)
 			return
 		}
 	}
+
+	if log.V(3) {
+		log.Infof("processing replica %s from %s queue...", repl, bq.name)
+	}
+	bq.eventLog.Printf("%s: processing", repl)
+
 	if err := bq.impl.process(now, repl, cfg); err != nil {
-		log.Errorf("while processing replica %s from %s queue: %s", repl, bq.name, err)
-	} else if log.V(2) {
-		log.Infof("processed replica %s from %s queue in %s", repl, bq.name, time.Now().Sub(start))
+		log.Errorf("while processing %s from %s queue: %s", repl, bq.name, err)
+		bq.eventLog.Printf("%s: error: %v", repl, err)
+	} else {
+		if log.V(2) {
+			log.Infof("processed %s from %s queue in %s", repl, bq.name, time.Now().Sub(start))
+		}
+		bq.eventLog.Printf("%s: done", repl)
 	}
 }
 

--- a/storage/store.go
+++ b/storage/store.go
@@ -338,10 +338,6 @@ type StoreContext struct {
 
 	// Tracer is a request tracer.
 	Tracer *tracer.Tracer
-
-	// ScannerStopper is used to shut down the background scanner (for tests).
-	// If nil, defaults to the store's own stopper.
-	ScannerStopper *stop.Stopper
 }
 
 // Valid returns true if the StoreContext is populated correctly.
@@ -565,15 +561,11 @@ func (s *Store) Start(stopper *stop.Stopper) error {
 		// Start the scanner. The construction here makes sure that the scanner
 		// only starts after Gossip has connected, and that it does not block Start
 		// from returning (as doing so might prevent Gossip from ever connecting).
-		scannerStopper := s.ctx.ScannerStopper
-		if scannerStopper == nil {
-			scannerStopper = s.stopper
-		}
-		scannerStopper.RunWorker(func() {
+		s.stopper.RunWorker(func() {
 			select {
 			case <-s.ctx.Gossip.Connected:
-				s.scanner.Start(s.ctx.Clock, scannerStopper)
-			case <-scannerStopper.ShouldStop():
+				s.scanner.Start(s.ctx.Clock, s.stopper)
+			case <-s.stopper.ShouldStop():
 				return
 			}
 		})

--- a/storage/store.go
+++ b/storage/store.go
@@ -439,6 +439,17 @@ func (s *Store) StartedAt() int64 {
 func (s *Store) Start(stopper *stop.Stopper) error {
 	s.stopper = stopper
 
+	// Add a closer for the various scanner queues, needed to properly clean up
+	// the event logs.
+	s.stopper.AddCloser(stop.CloserFn(func() {
+		s.gcQueue.Close()
+		s.splitQueue.Close()
+		s.verifyQueue.Close()
+		s.replicateQueue.Close()
+		s.replicaGCQueue.Close()
+		s.raftLogQueue.Close()
+	}))
+
 	if s.Ident.NodeID == 0 {
 		// Open engine (i.e. initialize RocksDB database). "NodeID != 0"
 		// implies the engine has already been opened.

